### ISSUE TITLE
codecov.io go coverage different than standard go coverage tool

### ISSUE THREAD

**Issue by vadmeste** [codecov/support#278]
Coverage rates seen in codecov.io are little different than the one generated by go tool cover -html. Is this intended?

**Comment by vadmeste** [comment]
I thought -X fix would be enough to have the same results but it wasn't the case

**Comment by stevepeak** [comment]
@vadmeste these changes are intended. Codecov works to create a very consistent user experience across all languages. Go-lang will measure empty lines in coverage which is a bad practice, Codecov fixes that.

**Comment by stevepeak** [comment]
The `-X fix` would be the way to ignore Codecov fixes. Can you provide an owner/repo@sha of a build that included `-X fix` in the upload?

**Comment by vadmeste** [comment]
@stevepeak, actually this is still a PR and not merged yet.

https://github.com/minio/minio/pull/2409
https://codecov.io/gh/minio/minio/pull/2409?src=pr

With this particular file (access-key.go) https://codecov.io/gh/minio/minio/src/91c09e0d750426539d11c044f24edcfe6b43ab16/access-key.go, `go coverage` says it has 81% but it has 63% in `codecov.io`

**Comment by stevepeak** [comment]
What does the Go coverage html look like for this particular file?

**Comment by vadmeste** [comment]
![go-coverage-access-key](https://cloud.githubusercontent.com/assets/283197/17629030/39d09396-60b0-11e6-88b2-d7331d0d3288.png)

**Comment by stevepeak** [comment]
What about the actual lines below?

**Comment by vadmeste** [comment]
yes sorry.. here it is.. by the way, I am using go 1.6.3

![screencapture-file-tmp-cover488413567-coverage-html-1471018981535](https://cloud.githubusercontent.com/assets/283197/17629303/8fa08078-60b1-11e6-8ac5-271548665580.png)

**Comment by stevepeak** [comment]
So the math on Codecov side is correct: 33 lines. 21 hit == 21/33 == 63%.

I would suggest using Codecov fixes because it would remove coverage data on the empty lines and bracket only lines.

**Comment by vadmeste** [comment]
Hello @stevepeak,

I am not sure that I understand your point. Isn't `-X fix` supposed to give the same coverage rate as the standard golang does ?

**Comment by stevepeak** [comment]
@vadmeste yes Codecov does report different metrics because it removes empty lines and ending brackets. `-X fix` would not ignore the prior. However, the actual calculation for coverage may be different because of the partial line coverage Go uses. 

The question becomes which is more accurate: Codecov or Go. Depends on your perspective. I would save give each option a shot and choose the one that fits your needs best.

> PS We are adding Go partial line coverage support soon. Thanks!

**Comment by vadmeste** [comment]
> PS We are adding Go partial line coverage support soon. Thanks!

Nice.

So if I understood correctly, Codecov will give similar results to Golang when partial line coverage support will be added ?

Notice that golang developers will try to enhance code coverage in their local machines using standard go tools, it would be little uncomfortable to see that rate changed after committing the code in the repository.

**Comment by udryan10** [comment]
@stevepeak any update on partial line coverage? We too are seeing this and would like for our local test coverage to match up with what codecov is reporting.

**Comment by stevepeak** [comment]
Hey friends, I'm ready to reinvestigate Go reporting. We will be open sourcing our coverage processing agents soon and can work together to get this working 💯

**Comment by njasm** [comment]
@stevepeak  any news about this issue?

**Comment by stevepeak** [comment]
@sbougerel made a very nice example showing this off https://github.com/sbougerel/codecov-go-example
